runtime/src/impls.rs

```diff
@@ -16,38 +16,60 @@
 
 //! Some configurable implementations as associated type for the substrate runtime.
 
-use crate::{Runtime, Treasury};
-use cennznet_primitives::{traits::SimpleAssetSystem, types::Balance};
+use crate::{BlockPayoutInterval, EpochDuration, Rewards, Runtime, SessionsPerEra, Staking, Treasury};
+use cennznet_primitives::{
+	traits::SimpleAssetSystem,
+	types::{AccountId, Balance},
+};
+use crml_staking::{rewards::RunScheduledPayout, EraIndex};
 use frame_support::{
 	dispatch::DispatchResult,
 	traits::{Contains, ContainsLengthBound, Currency, Get, OnUnbalanced},
-	weights::{WeightToFeeCoefficient, WeightToFeeCoefficients, WeightToFeePolynomial},
+	weights::{Weight, WeightToFeeCoefficient, WeightToFeeCoefficients, WeightToFeePolynomial},
 };
 use prml_generic_asset::{NegativeImbalance, StakingAssetCurrency};
 use prml_support::{AssetIdAuthority, MultiCurrencyAccounting};
 use smallvec::smallvec;
-use sp_runtime::{traits::Convert, Perbill};
+use sp_runtime::Perbill;
 use sp_std::{marker::PhantomData, prelude::*};
 
-/// Struct that handles the conversion of Balance -> `u64`. This is used for staking's election
-/// calculation.
-pub struct CurrencyToVoteHandler;
+/// Runs scheduled payouts for the rewards module.
+pub struct ScheduledPayoutRunner<T: crml_staking::rewards::Trait>(PhantomData<T>);
+// The max. number of payouts we can perform in an era are given by:
+// let MAX_PAYOUT_CAPACITY = (SessionsPerEra::get() * EpochDuration::get() / BlockPayoutInterval::get();
+// const_assert!(MAX_PAYOUT_CAPACITY >= Staking::maximum_validator_count())
+// NOTE: The # of validators elected should be less than this number or it will cause capacity issues.
+// config as of v39 has the upper limit of this number at: 5,760 elected validators.
+const MAX_PAYOUT_CAPACITY: u32 = SessionsPerEra::get() * EpochDuration::get() as u32 / BlockPayoutInterval::get();
+const MAX_VALIDATORS: u32 = 5_000;
+/// does the current config provide enough capacity to do automatic payouts
+/// failure here means a bad config or a new reward scaling solution should be sought if validator count is expected to be > 5_000
+static_assertions::const_assert!(MAX_PAYOUT_CAPACITY > MAX_VALIDATORS);
+
+impl<T: crml_staking::rewards::Trait> RunScheduledPayout for ScheduledPayoutRunner<T> {
+	type AccountId = AccountId;
+	type Balance = Balance;
 
-impl CurrencyToVoteHandler {
-	fn factor() -> Balance {
-		(<StakingAssetCurrency<Runtime>>::total_issuance() / u64::max_value() as Balance).max(1)
-	}
-}
+	/// Feed exposure info from staking to run reward calculations
+	/// This is called by Rewards on_initialize at scheduled intervals
+	fn run_payout(validator_stash: &Self::AccountId, amount: Self::Balance, payout_era: EraIndex) -> Weight {
```

Comment:
Instead of adding on_initialize to the rewards module, it could be added to the staking module. Then there staking module could interact with Rewards internally. With such a design, there wouldn't be any need to bring the implementation of staking/rewards functions to runtime.

Reply:
payout scheduling is something the staking module doesn't need to know about.
It belongs more in the rewards module but needed to pass a trait to share the staking exposure storage,
this would create a dependency cycle from staking needing 'T:Rewarder' and rewards needing 'T::LastExposureInfo', this design would've compiled fine but in the end decided to avoid the cycle and
use a higher-level component with knowledge of both modules